File: src/ipc/messages/inspect.ts
import * as valuetype from './valuetype';
import type { LeapStruct, LeapEnum, LeapVariant } from './valuetype';

export const leapStructsMap = new Map<string, LeapStruct> ([
    ["aggregate-entry", new valuetype.LeapStruct("aggregate-entry", [], [new valuetype.Property("global-id", new valuetype.ValueType("int", [])), new valuetype.Property("local-id", new valuetype.ValueType("int", [])), new valuetype.Property("name", new valuetype.ValueType("str", [])), new valuetype.Property("name-hash", new valuetype.ValueType("int", [])), new valuetype.Property("path", new valuetype.ValueType("str", [])), new valuetype.Property("self-size", new valuetype.ValueType("int", [])), new valuetype.Property("size", new valuetype.ValueType("int", [])), new valuetype.Property("tail-size", new valuetype.ValueType("int", [])), new valuetype.Property("self-file-count", new valuetype.ValueType("int", [])), new valuetype.Property("file-count", new valuetype.ValueType("int", [])), new valuetype.Property("tail-file-count", new valuetype.ValueType("int", [])), new valuetype.Property("self-dir-count", new valuetype.ValueType("int", [])), new valuetype.Property("dir-count", new valuetype.ValueType("int", [])), new valuetype.Property("tail-dir-count", new valuetype.ValueType("int", [])), new valuetype.Property("is-file", new valuetype.ValueType("bool", [])), new valuetype.Property("global-parent", new valuetype.ValueType("option", [new valuetype.ValueType("int", [])])), new valuetype.Property("local-parent", new valuetype.ValueType("option", [new valuetype.ValueType("int", [])])), new valuetype.Property("nested", new valuetype.ValueType("list", [new valuetype.ValueType("int", [])]))])],
    ["path-component", new valuetype.LeapStruct("path-component", [], [new valuetype.Property("path", new valuetype.ValueType("str", [])), new valuetype.Property("name", new valuetype.ValueType("str", []))])],
    ["aggregate-data", new valuetype.LeapStruct("aggregate-data", [], [new valuetype.Property("path", new valuetype.ValueType("str", [])), new valuetype.Property("path-top", new valuetype.ValueType("str", [])), new valuetype.Property("path-components", new valuetype.ValueType("list", [new valuetype.ValueType("path-component", [])])), new valuetype.Property("path-separator", new valuetype.ValueType("str", [])), new valuetype.Property("entries", new valuetype.ValueType("list", [new valuetype.ValueType("aggregate-entry", [])]))])],
    ["scan-progress", new valuetype.LeapStruct("scan-progress", [], [new valuetype.Property("done-count", new valuetype.ValueType("int", []))])],
    ["navigation", new valuetype.LeapStruct("navigation", [], [new valuetype.Property("global-id", new valuetype.ValueType("option", [new valuetype.ValueType("int", [])])), new valuetype.Property("path", new valuetype.ValueType("str", []))])],
    ["about-app", new valuetype.LeapStruct("about-app", [], [new valuetype.Property("version", new valuetype.ValueType("str", []))])],
    ["none", new valuetype.LeapStruct("none", [], [])],
    ["some", new valuetype.LeapStruct("some", ["t"], [new valuetype.Property("value", new valuetype.ValueType("t", []))])],
]);

export const leapEnumsMap = new Map<string, LeapEnum> ([
    ["scan-state", new valuetype.LeapEnum("scan-state", [], [new valuetype.Property("ready", new valuetype.ValueType("none", [])), new valuetype.Property("scan-progress", new valuetype.ValueType("scan-progress", []))])],
    ["option", new valuetype.LeapEnum("option", ["t"], [new valuetype.Property("none", new valuetype.ValueType("none", [])), new valuetype.Property("some", new valuetype.ValueType("some", [new valuetype.ValueType("t", [])]))])],
    ["result", new valuetype.LeapEnum("result", ["t", "e"], [new valuetype.Property("ok", new valuetype.ValueType("some", [new valuetype.ValueType("t", [])])), new valuetype.Property("err", new valuetype.ValueType("some", [new valuetype.ValueType("e", [])]))])],
]);

export const leapVariantsMap = new Map<string, Map<string, LeapVariant>>([
    [
        "scan-state",
        new Map<string, LeapVariant>([
            ["ready", new valuetype.LeapVariant("scan-state", new valuetype.Property("ready", new valuetype.ValueType("none", [])), [], [])],
            ["scan-progress", new valuetype.LeapVariant("scan-state", new valuetype.Property("scan-progress", new valuetype.ValueType("scan-progress", [])), [], [new valuetype.Property("done-count", new valuetype.ValueType("int", []))])],
        ])
    ],
    [
        "option",
        new Map<string, LeapVariant>([
            ["none", new valuetype.LeapVariant("option", new valuetype.Property("none", new valuetype.ValueType("none", [])), ["t"], [])],
            ["some", new valuetype.LeapVariant("option", new valuetype.Property("some", new valuetype.ValueType("some", [new valuetype.ValueType("t", [])])), ["t"], [new valuetype.Property("value", new valuetype.ValueType("t", []))])],
        ])
    ],
    [
        "result",
        new Map<string, LeapVariant>([
            ["ok", new valuetype.LeapVariant("result", new valuetype.Property("ok", new valuetype.ValueType("some", [new valuetype.ValueType("t", [])])), ["t", "e"], [new valuetype.Property("value", new valuetype.ValueType("t", []))])],
            ["err", new valuetype.LeapVariant("result", new valuetype.Property("err", new valuetype.ValueType("some", [new valuetype.ValueType("e", [])])), ["t", "e"], [new valuetype.Property("value", new valuetype.ValueType("e", []))])],
        ])
    ],
]);

File: src-tauri/src/ipc/messages/inspect.rs
use std::collections::HashMap;
use super::value_type;

pub fn get_leap_structs_map() -> HashMap<String, value_type::LeapStruct> {
    [
        ("aggregate-entry".to_owned(), value_type::LeapStruct::new("aggregate-entry".to_owned(), vec![], vec![value_type::Property::new("global-id".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("local-id".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("name".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("name-hash".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("path".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("self-size".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("size".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("tail-size".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("self-file-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("file-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("tail-file-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("self-dir-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("dir-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("tail-dir-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![])), value_type::Property::new("is-file".to_owned(), value_type::ValueType::new("bool".to_owned(), vec![])), value_type::Property::new("global-parent".to_owned(), value_type::ValueType::new("option".to_owned(), vec![value_type::ValueType::new("int".to_owned(), vec![])])), value_type::Property::new("local-parent".to_owned(), value_type::ValueType::new("option".to_owned(), vec![value_type::ValueType::new("int".to_owned(), vec![])])), value_type::Property::new("nested".to_owned(), value_type::ValueType::new("list".to_owned(), vec![value_type::ValueType::new("int".to_owned(), vec![])]))])),
        ("path-component".to_owned(), value_type::LeapStruct::new("path-component".to_owned(), vec![], vec![value_type::Property::new("path".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("name".to_owned(), value_type::ValueType::new("str".to_owned(), vec![]))])),
        ("aggregate-data".to_owned(), value_type::LeapStruct::new("aggregate-data".to_owned(), vec![], vec![value_type::Property::new("path".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("path-top".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("path-components".to_owned(), value_type::ValueType::new("list".to_owned(), vec![value_type::ValueType::new("path-component".to_owned(), vec![])])), value_type::Property::new("path-separator".to_owned(), value_type::ValueType::new("str".to_owned(), vec![])), value_type::Property::new("entries".to_owned(), value_type::ValueType::new("list".to_owned(), vec![value_type::ValueType::new("aggregate-entry".to_owned(), vec![])]))])),
        ("scan-progress".to_owned(), value_type::LeapStruct::new("scan-progress".to_owned(), vec![], vec![value_type::Property::new("done-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![]))])),
        ("navigation".to_owned(), value_type::LeapStruct::new("navigation".to_owned(), vec![], vec![value_type::Property::new("global-id".to_owned(), value_type::ValueType::new("option".to_owned(), vec![value_type::ValueType::new("int".to_owned(), vec![])])), value_type::Property::new("path".to_owned(), value_type::ValueType::new("str".to_owned(), vec![]))])),
        ("about-app".to_owned(), value_type::LeapStruct::new("about-app".to_owned(), vec![], vec![value_type::Property::new("version".to_owned(), value_type::ValueType::new("str".to_owned(), vec![]))])),
        ("none".to_owned(), value_type::LeapStruct::new("none".to_owned(), vec![], vec![])),
        ("some".to_owned(), value_type::LeapStruct::new("some".to_owned(), vec!["t".to_owned()], vec![value_type::Property::new("value".to_owned(), value_type::ValueType::new("t".to_owned(), vec![]))])),
    ].into()
}

pub fn get_leap_enums_map() -> HashMap<String, value_type::LeapEnum> {
    [
        ("scan-state".to_owned(), value_type::LeapEnum::new("scan-state".to_owned(), vec![], vec![value_type::Property::new("ready".to_owned(), value_type::ValueType::new("none".to_owned(), vec![])), value_type::Property::new("scan-progress".to_owned(), value_type::ValueType::new("scan-progress".to_owned(), vec![]))])),
        ("option".to_owned(), value_type::LeapEnum::new("option".to_owned(), vec!["t".to_owned()], vec![value_type::Property::new("none".to_owned(), value_type::ValueType::new("none".to_owned(), vec![])), value_type::Property::new("some".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("t".to_owned(), vec![])]))])),
        ("result".to_owned(), value_type::LeapEnum::new("result".to_owned(), vec!["t".to_owned(), "e".to_owned()], vec![value_type::Property::new("ok".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("t".to_owned(), vec![])])), value_type::Property::new("err".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("e".to_owned(), vec![])]))])),
    ].into()
}

pub fn get_leap_variants_map() -> HashMap<String, HashMap<String, value_type::LeapVariant>> {
    [
        (
            "scan-state".to_owned(),
            [
                ("ready".to_owned(), value_type::LeapVariant::new("scan-state".to_owned(), value_type::Property::new("ready".to_owned(), value_type::ValueType::new("none".to_owned(), vec![])), vec![], vec![])),
                ("scan-progress".to_owned(), value_type::LeapVariant::new("scan-state".to_owned(), value_type::Property::new("scan-progress".to_owned(), value_type::ValueType::new("scan-progress".to_owned(), vec![])), vec![], vec![value_type::Property::new("done-count".to_owned(), value_type::ValueType::new("int".to_owned(), vec![]))])),
            ].into()
        ),
        (
            "option".to_owned(),
            [
                ("none".to_owned(), value_type::LeapVariant::new("option".to_owned(), value_type::Property::new("none".to_owned(), value_type::ValueType::new("none".to_owned(), vec![])), vec!["t".to_owned()], vec![])),
                ("some".to_owned(), value_type::LeapVariant::new("option".to_owned(), value_type::Property::new("some".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("t".to_owned(), vec![])])), vec!["t".to_owned()], vec![value_type::Property::new("value".to_owned(), value_type::ValueType::new("t".to_owned(), vec![]))])),
            ].into()
        ),
        (
            "result".to_owned(),
            [
                ("ok".to_owned(), value_type::LeapVariant::new("result".to_owned(), value_type::Property::new("ok".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("t".to_owned(), vec![])])), vec!["t".to_owned(), "e".to_owned()], vec![value_type::Property::new("value".to_owned(), value_type::ValueType::new("t".to_owned(), vec![]))])),
                ("err".to_owned(), value_type::LeapVariant::new("result".to_owned(), value_type::Property::new("err".to_owned(), value_type::ValueType::new("some".to_owned(), vec![value_type::ValueType::new("e".to_owned(), vec![])])), vec!["t".to_owned(), "e".to_owned()], vec![value_type::Property::new("value".to_owned(), value_type::ValueType::new("e".to_owned(), vec![]))])),
            ].into()
        ),
    ].into()
}
